[DependencyInjection] Fluent PHP DI Documentation

ragboyjr opened the pull request:
- Updated service_container.rst to include php-fluent-di
- Updated _build/conf.py to include codeblock for php-fluent-di
- Updated existing di resource loading to refer to latest sf 4.2
  excludes

Remaining: need to update the rest of the service_container files

Signed-off-by: RJ Garcia <rj@bighead.net>

ragboyjr committed:
[DependencyInjection] Fluent PHP DI Documentation

- Updated DI configuration references to include php-fluent-di
- Updated _build/conf.py to include codeblock for php-fluent-di
- Updated existing di resource loading to refer to latest sf 4.2
  excludes

Signed-off-by: RJ Garcia <rj@bighead.net>

ragboyjr commented:
Before I start moving on to the rest of the files, would be nice if I could get some feedback to make sure this approach is fine and will work across all of our documentation.

ragboyjr commented:
Any word on this? Willing to make adjustments, just want to make sure i'm not wasting anyones efforts.

ragboyjr commented:
@nicolas-grekas any thoughts?

nicolas-grekas reviewed (approved):
LGTM, thanks for the ping and the PR.

ragboyjr commented:
K, 4 files left now:

```
service_container/
  service_decoration.rst
  service_subscribers_locators.rst
  shared.rst
  tags.rst
```

Might be able to finish the rest tomorrow.

ragboyjr commented:
@nicolas-grekas OK, all of the files have been updated, and I was able to build the site locally, and it all checked out. This is ready to go live IMHO.

ragboyjr commented:
@nicolas-grekas is there anything else i need to do to get this merged? Would love to start seeing the fluent php configuration show up on the site so I can link to it for the devs i work with.

ragboyjr commented:
oh, ha, no worries! Thank you very much.

ragboyjr commented:
@javiereguiluz is there anything else I need to do here?

javiereguiluz commented:
@ragboyjr you did everything great and I thank you for that! However, merging this is a bit complicated. Why? Because we already have PHP config in all examples and here we are adding ... another PHP config.

Showing two similar, but slightly different, configs is always confusing to users. Why two? Which ons is the right one? What if I choose one and the right one is the other one? etc. Instead of showing both, I think we should replace the existing PHP config by the new PHP config. However, is this fluent config polished enough? Is it used enough? I don't hear people asking questions about this or saying that they use it.

I'm going to ask about this to the Docs Team and we'll be back with a reply. Thanks!

ragboyjr commented:
@javiereguiluz I appreciate the feedback!

My only two cents is that the PHP Fluent config won't ever get "used" or "polished" enough unless it's documented properly. A lot of that information in the docs I just got from using it, IDE auto-complete, and checking source code, and I've found that it's hard to sell a team on trying to use this format when there isn't any documentation on it except for a blog post by the symfony team for the 3.4 release.

All that to say, I think before it could ever replace the other PHP format, it would need to at least be documented so that people could use it, and we can work out the kinks.

I totally understand your concern regarding showing both php types would be confusing; however, I'm not sure dropping the default php version would be good since it does provide valuable information on how to interact with the DI container. It's especially helpful when using Compiler Passes.

The only solutions I could come up with would be to change the name of it to not be `PHP (Fluent)`, but give it a name like `Configurator` or `Fluent` maybe. The idea being that it lets people know it's a totally different configuration file.

The other would be that we put this PR on the back burner, and we create a dedicated page show casing each of the individual parts of DI configuration using the PHP fluent format, and it could be apart of the learn more.

Then after some time, if we feel like PHP fluent is taking off etc etc, then we can add it into the main docs or something.

nicolas-grekas commented:
I'd be :+1: to fade out the old PHP format - it doesn't provide anything over the fluent one, except maybe low level access to primitive data structures.

javiereguiluz commented:
I did a quick survey on Symfony's Slack. The number of answers is small ... but the results were:

1) Almost nobody uses PHP for config and lots never heard of this fluent format.
2) Of those using PHP, most are using (and loving) this fluent format.

ragboyjr commented:
@javiereguiluz That actually fits with my experience as well.

ragboyjr commented:
@javiereguiluz Let me know what you want to do, willing to update this however you want. The hard part was just creating the mapped config. If I have to shuffle stuff around or remove some stuff, it'll be quick.

ragboyjr commented:
@javiereguiluz ping, here to help and contribute any you need/like. Would love to see the comprehensive fluent php docs show up on the SF docs site in any shape or form.

ragboyjr commented:
Understood, just checking in, thanks.

ragboyjr commented:
@javiereguiluz gentle ping. I know we're all busy, just bringing this back up to the top of your email :), so it doesn't slip into the abyss.

javiereguiluz commented:
I'm very sorry for the late reply! After discussing about this ... we agree on showing this new fluent format ... but not in addition to but replacing the old non-fluent PHP code.

We've already doing that in other parts of the docs, such as routes: #11085.

@ragboyjr would you be able to make these changes so we can merge it? Thanks a lot!

OskarStark commented on service_container/3.3-di-changes.rst:
```diff
@@ -81,9 +81,32 @@ what the file looks like in Symfony 4):
 
+.. versionadded:: 3.4
+
+    PHP Fluent DI configuration wasn't available until version 3.4.
```
PHP Fluent DI was introduced ja Symfony 3.4.

ragboyjr replied:
I'm not sure if you are agreeing with me or you need to make a change..?

OskarStark replied:
IMO my sentence is more like in the other sf docs.

What do you think @javiereguiluz ?

ragboyjr replied:
ah, I understand, you want me to change the content of my versionadded note to: `PHP Fluent DI was introduced in Symfony 3.4.`

ragboyjr replied:
sounds good to me.

ragboyjr replied:
Resolved.

OskarStark commented on service_container/alias_private.rst:
```diff
@@ -151,7 +176,14 @@ Anonymous Services
+
+    The feature to configure anonymous services in YAML was introduced in Symfony 3.3.
+
+.. versionadded:: 4.1
```
Please add a linebreak afterwards

ragboyjr replied:
Resolved.

HeahDude commented on components/dependency_injection.rst:
```diff
@@ -302,6 +302,25 @@ config files:
 
+    .. code-block:: php-fluent-di
+
+        use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
```
Should be `namespace Symfony\Component\DependencyInjection\Loader\Configurator;`

ragboyjr replied:
kk, will take a look.

HeahDude replied:
I think this is a basic and deserves an addition in all example :). What do others think?

ragboyjr replied:
OK, @HeahDude I just checked, and my documentation is right. 

`Symfony\Component\DependencyInjection\Loader\Configurator;` isn't an actual class, it's `ContianerConfigurator` like mine shows. Also - https://symfony.com/blog/new-in-symfony-3-4-php-based-configuration-for-services-and-routes shows ContainerConfigurator too.

And I also checked my 3 sites that use the fluent di configuration and they all use ContainerConfigurator as well.

HeahDude replied:
Maybe you misread my comment, I mean you should add the namespace and remove all use statements. It would work the same and would be cleaner. We do this for routing too.

ragboyjr replied:
Ah, I definitely did mis-understand your comment. 

So are you asking that I update every place where I import `use function ...` and replace an example function call to `ref('id')` with something like `Configurator\ref('id')`?

if so, how would you feel about doing something like:

```
use Symfony\Component\DependencyInjection\Loader\Configurator as di;

ragboyjr replied:
Also, for precedent, when importing symfony validation constraints, the standard is to alias `Symfony\Component\Validation\Constraints as Assert` to make the dx better for using the constraints.

HeahDude replied:
Using the namespace should prevent the need of any use statement:
```php
namespace Symfony\Component\DependencyInjection\Loader\Configurator;

return function(ContainerConfigurator $configurator) {
    // ...
    $container->set('svc')->args([ref('abc')];
};
```

ragboyjr replied:
Ai, ok, that makes sense, will update across the board.

HeahDude replied:
Thank you!

HeahDude commented on components/dependency_injection.rst:
```diff
@@ -302,6 +302,25 @@ config files:
+
+            $container->set('newsletter_manager', 'NewsletterManager')
+                ->call('setMailer', [ref('mailer')]);
+        };
```
What about showing an alias too?

ragboyjr replied:
up to you boss, i'll do whatever. I tried to keep the examples exactly the same as the yaml/xml counterparts to be consistent.

HeahDude replied:
I think this is a basic and deserves an addition in all example :). What do others think?

ragboyjr replied:
I updated every place that references di configuration, and we do have several examples that show using aliases. I think if we *were* to update the main examples to showcase aliases, that should be done in a separate as a separate change.

HeahDude reviewed (commented):
I did not review all the PR, but I would be in favour of using this instead of the current PHP config, as done for routes.

We should only talk about low level implementation when talking about compiler passes, or as simple example in the main component page.

ragboyjr commented:
@javiereguiluz Awesome, yes sir, will remove the PHP (Fluent DI) sections and the extra config for sphinx and just replace the php definitions.

I can probably turn something out tonight.

ragboyjr commented:
@javiereguiluz OK, I rebased off of the latest master, and made two more commits, one to replace the php config with fluent-di, and another to address the comments by @OskarStark. 

I can rebase/squash these commits once you guys review and approve of the changes.

Thanks again for the opportunity to contribute!

ragboyjr commented:
@HeahDude ok, i updated everything to use `namespace ...` and removed the `use` and `use function` imports. Can you take a look?

HeahDude commented on components/dependency_injection.rst:
```diff
@@ -290,17 +290,21 @@ config files:
+            $configurator->parameters()
+                ->set('mailer.transport', 'sendmail');
+
+            $container = $configurator->services();
```
I am not whether we should go for this or:
```php
return function (ContainerConfigurator $container) {
    $container->parameters()
        ->set('mailer.transport', 'sendmail')
        ->set('xxx', 'XXX')
    ;

    $container->services()
        ->set(...)->args(...)
        ->set(...)->call()->public()

        ->instancof()->tags()
        // ...
```
WDYT?
There are so many paths. Playing with this is a real joy, I guess I won't be able to use any other format now :).
I would like we all agree on a best practice for this to roll it out.
ping @symfony/team-symfony-docs @nicolas-grekas

ragboyjr replied:
K, so I actually have good use case for why you should keep the master `$configurator` separate from the services configurator named: `$container` .

There are times when extending an abstract service that you cannot be within the context of a service configurator with defaults set. So you need to keep a reference to the original ContainerConfigurator instance to do so.

ragboyjr replied:
@HeahDude not sure it matters to much, I think I personally prefer a separate statement per definition just so that it makes the definitions more distinct in the code, but people can also style their stuff however they want.

HeahDude commented on service_container.rst:
```diff
@@ -175,26 +175,25 @@ each time you ask for it.
+                    ->defaults()
+                    ->autowire()
+                    ->autoconfigure()
+                    ->private();
```
Private should be removed every where on 4.2, this could be done in a dedicated PR too.

ragboyjr replied:
Is that just because services are now private by default? If so, should we have documentation describing that services are private by default starting in 4.2?

HeahDude replied:
This is the case since Symfony 4. 4.2 is just now the lowest 4.x branch maintained.

HeahDude commented on service_container.rst:
```diff
@@ -175,26 +175,25 @@ each time you ask for it.
-            // To use as default template
-            $definition = new Definition();
+            return function(ContainerConfigurator $configurator) {
+                $container = $configurator->services()
```
Same here, what about:
```php
$container->services()
    ->defaults()->autowire()->autoconfigure()

    ->load('App\\', '../src/*')->exclude('../src/{DependencyInjection,Entity,Migrations,Tests,Kernel.php}')

    ->set('x')->arg('$arg', ref('...'))
    ->alias('y', 'x')->public()

    // ...
;
```
?

HeahDude commented on service_container.rst:
```diff
@@ -503,12 +502,19 @@ and reference it with the ``%parameter_name%`` syntax:
+
+            $container = $configurator->services();
+            $container->set(SiteUpdateManager::class)
+                ->autowire()
```
To be consistent with other examples you should add an `// ...` before and remove `->autowire()`

ragboyjr replied:
Good catch, i'll go through and make those more consistent, I thought I had done that, but I obviously missed some.

HeahDude commented on service_container.rst:
```diff
@@ -612,13 +618,18 @@ But, you can control this and pass in a different logger:
+        return function(ContainerConfigurator $configurator) {
+            $container = $configurator->services();
+            $container->set(SiteUpdateManager::class)
+                ->autowire()
```
Can you please add the same comments as other examples and remove needless boilerplate?

ragboyjr replied:
Y, definitely, not sure why I did it like that.

HeahDude commented on service_container.rst:
```diff
@@ -700,21 +711,21 @@ You can also use the ``bind`` keyword to bind specific arguments by name or type
+                ->bind('$adminEmail', 'manager@example.com')
+                ->bind('$requestLogger', ref('monolog.logger.request'))
+                ->bind(LoggerInterface::class, ref('monolog.logger.request'))
+                ->bind('string $adminEmail', 'manager@example.com')
```
Please don't remove comments

HeahDude commented on service_container.rst:
```diff
@@ -1038,27 +1059,31 @@ admin email. In this case, each needs to have a unique service id:
+        return function(ContainerConfigurator $configurator) {
+            // ...
+
+            $container->set('site_update_manager.superadmin', SiteUpdateManager::class)
```
```php
$container->services()
    // ...
    
    ->set('site_update_manager.superadmin', SiteUpdateManager::class)->autowire(false)->args([
        ref(MessageGenerator::class),
        ref('mailer'),
        'superadmin@example.com',
    ])

    ->set(...)
```
?

HeahDude commented on service_container/3.3-di-changes.rst:
```diff
@@ -81,9 +81,32 @@ what the file looks like in Symfony 4):
 This small bit of configuration contains a paradigm shift of how services
 are configured in Symfony.
 
+.. versionadded:: 3.4
```
This can be removed while targeting 4.2 :).

HeahDude commented on service_container/3.3-di-changes.rst:
```diff
@@ -462,6 +501,21 @@ inherited from an abstract definition:
+        return function(ContainerConfigurator $configurator) {
+            // ...
+
+            $container->instanceof(LoaderInterface::class
```
Missing parenthesis?

I would go for a :
```php
$container->services()
    // ...

    ->instanceof(LoaderInterface::class)->tag('app.domain_loader')->public()
;
```

HeahDude commented on service_container/alias_private.rst:
```diff
@@ -55,10 +55,15 @@ You can also control the ``public`` option on a service-by-service basis:
+        return function(ContainerConfigurator $configurator) {
+            $container = $configurator->services();
+            $container->set(Foo::class)
+                ->private();
```
Can be removed.

HeahDude commented on service_container/alias_private.rst:
```diff
@@ -214,7 +223,15 @@ Anonymous Services
-    Anonymous services are only supported by the XML and YAML configuration formats.
+    Anonymous services are only supported by the XML, YAML, and PHP Fluent configuration formats.
+
+.. versionadded:: 3.3
```
Can be removed as well.

HeahDude commented on service_container/alias_private.rst:
```diff
@@ -214,7 +223,15 @@ Anonymous Services
 .. note::
 
-    Anonymous services are only supported by the XML and YAML configuration formats.
+    Anonymous services are only supported by the XML, YAML, and PHP Fluent configuration formats.
```
```diff
- PHP Fluent
+ :class:`Symfony\\Component\\DependencyInjection\\Loader\\Configurator\\ContainerConfigurator`
```

HeahDude commented on service_container/alias_private.rst:
```diff
@@ -214,7 +223,15 @@ Anonymous Services
+
+.. versionadded:: 4.1
+
+    The feaature to configure anonymous services in PHP Fluent was introduced in Symfony 3.4.
```
Can be reverted.

HeahDude commented on service_container/alias_private.rst:
```diff
@@ -252,6 +269,26 @@ The following example shows how to inject an anonymous service into another serv
+.. note::
+
+    Anonymous services do *NOT* inherit the definitions provided from the defaults defined in the configuration. So you'll
+    need to explicitly mark service as autowired or autoconfigured when doing an anonymous service e.g.: `inline(Foo::class)->autowire()->autoconfigure()`.
```
We use double back ticks for inline formatting:
```diff
- `inline(Foo::class)->autowire()->autoconfigure()`
+ ``inline(Foo::class)->autowire()->autoconfigure()``
```

ragboyjr replied:
kk

HeahDude commented on service_container/autowiring.rst:
```diff
@@ -229,13 +231,19 @@ adding a service alias:
+
+            $container->set('app.rot13.transformer', Rot13Transformer::class)
+                ->autowire()
+                ->private();
```
Can be removed.

HeahDude commented on service_container/autowiring.rst:
```diff
@@ -104,17 +104,19 @@ both services:
+                ->defaults()
+                ->autowire()
+                ->autoconfigure()
+                ->private();
```
Can be removed.

HeahDude commented on service_container/injection_types.rst:
```diff
@@ -228,12 +235,15 @@ Another possibility is setting public fields of the class directly::
+        return function(ContainerConfigurator $configurator) {
+            $container = $configurator->services();
+            $container->set('app.newsletter_manager, NewsletterManager::class)
+                ->property('mailer', ref('mailer'));
```
We should maybe show both `->arg('$mailer', ref('mailer))` and this with a comment.

HeahDude commented on service_container/parameters.rst:
```diff
@@ -144,11 +155,13 @@ and hidden with the service definition:
-            $container->setParameter('mailer_password', '@securepass');
+            namespace Symfony\Component\DependencyInjection\Loader\Configurator;
 
-            // But % does need to be escaped
```
Can you please revert those changes?

HeahDude commented on service_container/parent_services.rst:
```diff
@@ -117,28 +117,21 @@ duplicated service definitions:
-
-        // ...
+        return function(ContainerConfigurator $configurator) {
+            $configurator->services()
```
```php
$container->services()
    // NOTE: this must be set before a "defaults()" call so that it can be extended
    ->set(BaseDoctrineRepository::class)->abstract()
    ->args([ref('doctrine.orm.entity_manager')])
    ->call('setLogger', [ref('logger')->nullOnInvalid()])

    ->set(DoctrineUserRepository::class)->parent(BaseDoctrineRepository::class)
    ->set(DoctrinePostRepository::class)->parent(BaseDoctrineRepository::class)

    // ...
;
```
?

HeahDude commented on service_container/tags.rst:
```diff
@@ -38,11 +38,17 @@ example:
+        return function(ContainerConfigurator $configurator) {
+            $container = $configurator->services();
+            $container->set(AppExtension::class)
+                ->private()
```
Can be removed.

HeahDude commented on service_container/tags.rst:
```diff
@@ -500,17 +526,20 @@ first  constructor argument to the ``App\HandlerCollection`` service:
-            ->addTag('app.handler');
+        return function(ContainerConfigurator $configurator) {
+            $container = $configurator->services();
+            $container->set(App\Handler\One::class)
```
What about some use statements?

HeahDude reviewed (changes requested):
That looks really good, thank you for working on this. I did not comment everywhere about the style, most of the comments must be discussed before further changes IMHO, some are mandatory for 4.2 and consistency between formats which is hard to keep.

Thanks!

ragboyjr commented on components/dependency_injection.rst:
```diff
@@ -290,17 +290,21 @@ config files:
+
+        return function(ContainerConfigurator $configurator) {
+            $configurator->parameters()
+                ->set('mailer.transport', 'sendmail');
```
semi-colon on newline feels weird to me personally, but the reduced diff argument is a good one. I'd imagine a separate PR to clean that up and make consistent would be advised.

ragboyjr commented:
k, @HeahDude to sum up what I need to fix:

1. Remove calls to `private`
2. Fix the random issues you pointed out like, missing paren, double backtick instead of single
3. Update all php examples to match the comments and skip boilerplate exactly like the yaml/xml examples do.
4. Potentially update all formatting to basically be one entire expression instead of multiple statements.

I can do item 4 in this list, but I'd just need a firm answer on if we should go that route or not.

Personally, I don't care too much, but in medium sized configuration files, having spaces and comments between definitions is very helpful (and might feel a bit weird to do that in the expression version you show), and in even bigger configuration files, I'd do something like:

```php
return new class() {
    public function __invoke(ContainerConfigurator $configurator) {
        // setup the defaults
        $container = $configurator->defaults(); //...
        $this->module1($container, $configurator);
        $this->module2($container, $configurator);
        $this->module3($container, $configurator);
    }

    private function module1(AbstractConfigurator $container, ContainerConfigurator $configurator) {
        // register services
    }

    private function module2(AbstractConfigurator $container, ContainerConfigurator $configurator) {
        // register services
    }

    private function module3(AbstractConfigurator $container, ContainerConfigurator $configurator) {
        // register services
    }
};
```

ragboyjr commented:
@javiereguiluz @nicolas-grekas can either of you give some consensus on @HeahDude's comments regarding syntax for the configuration?

We currently do:

```php
return function(ContainerConfigurator $configurator) {
    $container = $configuator->services()->default()->autowire()->autoconfigure();
    
    $container->set()->factory();
    $container->set()->args()
    // ... etc etc
};
```

@HeahDude suggested something along the lines of:

```php
return function(ContainerConfigurator $container) {
    $container->services()->default()->autowire()->autoconfigure()
        ->set()->factory()
        ->set()->args()
        // ... etc etc
};
```

ragboyjr commented:
@HeahDude how about I just follow the conventions we did for the routing config, that's one statement per definition, and the `;` goes on a separate line. So if we want to change the style like you mentioned it, we can do the DI and routing config at the same time.

ragboyjr commented:
Gentle ping @HeahDude @javiereguiluz @nicolas-grekas

javiereguiluz commented:
@xabbuh @OskarStark @wouterj after reviewing this ... I'm afraid I cannot merge it. I never use this PHP format, so I can't fix any of the multiple conflicts that will happen when upmerging this to 4.3, 4.4 and master.

Would any of you available for this? Keep in mind that the merging will be very complex and time-consuming. Thanks!

ragboyjr commented:
@javiereguiluz @xabbuh @wouterj Let me see if I can resolve the merge conflicts.

javiereguiluz commented:
@ragboyjr thanks a lot for all your work in this pull request. I hope it can be merged sooner than later.

ragboyjr commented:
@javiereguiluz OK, check now please sir. I rebased off of master.

@HeahDude I updated several of the references to use the `;` on a new line, but i'm not sure all our switched over. I'm not sure it's worth holding up the PR at this point, we could probably have someone comb through and adjust the CS in a separate PR.

ragboyjr commented:
gah, @javiereguiluz I rebased off of master and not 3.4 :(. When rebasing off of 3.4 i get another set of conflicts :'(. Are we sure we want to target this documentation for 3.4? Or maybe a later branch?

javiereguiluz commented:
@ragboyjr you've spent so much time here that I feel bad if I keep asking things .... but yes, adding this only to 4.3 and higher branches would help a lot when merging it.

javiereguiluz commented:
Yes, please. This should go either to 3.4 or 4.3 .. so let's try if 4.3 is easier for all. Thanks!

ragboyjr commented:
Rebased off of 4.3 and branch target is updated.

wouterj commented:
Hi @ragboyjr! What an incredible job did you do with this PR. I cannot imagine the effort you put into this.

That's makes it even worse that we ignored it for so long. Fortunately, it's now merged into 4.3/4.4/5.0. While merging, I fixed some issues found by our review tool in 3ccce23 and I checked all comments by @HeahDude and applied some of them - as well as some things I discovered - in 2a5b114.

Thanks a lot for your quick responses and actions this week!

ragboyjr commented:
@wouterj @javiereguiluz woot woot!!!! So excited to see this in action now!

Glad to be able to help out with this :).
